### PR TITLE
implement packed struct equality

### LINKED ISSUE #21678: allow equality operator for packed structs
```zig
const std = @import("std");
const expect = std.testing.expect;

test "packed struct equality" {
    const S = packed struct {
        a: u4,
        b: u4,
    };
    const x: S = .{ .a = 1, .b = 2 };
    const y: S = .{ .b = 2, .a = 1 };
    try expect(x == y);
}
```

```
test.zig:11:18: error: operator == not allowed for type 'test.test.packed struct equality.S'
    try expect(x == y);
               ~~^~~~
test.zig:5:22: note: struct declared here
    const S = packed struct {
              ~~~~~~~^~~~~~
```

This test should pass, for the same reason that equality is allowed for enums.

### PR DESCRIPTION
**Pull request by Rexicon226**
closes #21678

### CODE REVIEW

File: test/behavior/packed_struct_equality.zig [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+test "packed struct equality" {
+    const S = packed struct {
+        a: u4,
+        b: u4,
+    };
+    const x: S = .{ .a = 1, .b = 2 };
+    const y: S = .{ .b = 2, .a = 1 };
+    try expect(x == y);
+    try expect(!(x != y));
```
Comment (andrewrk) [inline test/behavior/packed_struct_equality.zig:12]: please also test runtime

**Review by andrewrk — CHANGES_REQUESTED** [review]
Nice!

File: test/behavior/packed_struct_equality.zig (hunk 1 above)
Comment (Rexicon226) [reply test/behavior/packed_struct_equality.zig:12]: yep, good point. forgot.

Comment (Rexicon226) [reply test/behavior/packed_struct_equality.zig]: done, and split into two clearer commits.

**Comment by rohlem** [comment]
I don't see an explicit mention (nor test) of this, but IMO `a == b` should still be a compile error if `@TypeOf(a) != @TypeOf(b)`.
I assume the unchanged parts of the code already provide this behavior?

**Comment by Rexicon226** [comment]
> I don't see an explicit mention (nor test) of this, but IMO `a == b` should still be a compile error if `@TypeOf(a) != @TypeOf(b)`. I assume the unchanged parts of the code already provide this behavior?

It's a compile error because comparisons are peer resolves, and two different packed structs types cannot coerce into each other.

File: test/behavior/packed_struct_equality.zig [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    const S = struct {
+        fn doTest(x: Foo, y: Foo) !void {
+            try expect(x == y);
+            try expect(!(x != y));
+        }
+    };
+
+    const x: S = .{ .a = 1, .b = 2 };
+    const y: S = .{ .b = 2, .a = 1 };
```
Comment (PauloCampana) [inline test/behavior/packed_struct_equality.zig:18]: x and y are of type S, but S has no fields?

Comment (mlugg) [reply test/behavior/packed_struct_equality.zig:18]: Yeah -- @Rexicon226, you forgot to import this new file into `behavior.zig`. I'd suggest just putting it into an existing file, like `structs.zig` (`struct.zig`? can't check rn)

Comment (Vexu) [reply test/behavior/packed_struct_equality.zig:18]: Or `packed-struct.zig`.

Comment (Rexicon226) [reply test/behavior/packed_struct_equality.zig]: Thanks for noticing, I always forget that the behavior tests aren't auto-gathered like the compile error ones, moved to `packed-struct.zig`.

**Comment by Rexicon226** [comment]
@andrewrk will you write the release notes or should I?

**Comment by andrewrk** [comment]
> @andrewrk will you write the release notes or should I?

I will if you don't. You're certainly welcome to lighten my load :-)

**Pull request merged**